Remotion/Data/DomainObjects/Mapping/InlineTypeDefinitionVisitor.cs

```diff
@@ -0,0 +1,118 @@
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Mapping
+{
+  /// <summary>
+  /// Visits the given <see cref="TypeDefinition"/> and executes a handler based on the entity's type.
+  /// </summary>
+  public static class InlineTypeDefinitionVisitor
+  {
+    private class TypeDefinitionVisitor<T> : ITypeDefinitionVisitor<T>
+    {
+      private readonly Func<ITypeDefinitionVisitor<T>, ClassDefinition, T?> _classDefinitionHandler;
+      private readonly Func<ITypeDefinitionVisitor<T>, InterfaceDefinition, T?> _interfaceDefinitionHandler;
+
+      public TypeDefinitionVisitor (
+          Func<ITypeDefinitionVisitor<T>, ClassDefinition, T?> classDefinitionHandler,
+          Func<ITypeDefinitionVisitor<T>, InterfaceDefinition, T?> interfaceDefinitionHandler)
+      {
+        ArgumentUtility.CheckNotNull(nameof(classDefinitionHandler), classDefinitionHandler);
+        ArgumentUtility.CheckNotNull(nameof(interfaceDefinitionHandler), interfaceDefinitionHandler);
+
+        _classDefinitionHandler = classDefinitionHandler;
+        _interfaceDefinitionHandler = interfaceDefinitionHandler;
+      }
+
+      /// <inheritdoc />
+      public T? VisitClassDefinition (ClassDefinition classDefinition)
+      {
+        ArgumentUtility.CheckNotNull("classDefinition", classDefinition);
+
+        return _classDefinitionHandler(this, classDefinition);
+      }
+
+      public T? VisitInterfaceDefinition (InterfaceDefinition interfaceDefinition)
+      {
+        ArgumentUtility.CheckNotNull("interfaceDefinition", interfaceDefinition);
+
+        return _interfaceDefinitionHandler(this, interfaceDefinition);
+      }
+    }
+
+    private class TypeDefinitionVisitor : ITypeDefinitionVisitor
+    {
+      private readonly Action<ITypeDefinitionVisitor, ClassDefinition> _classDefinitionHandler;
+      private readonly Action<ITypeDefinitionVisitor, InterfaceDefinition> _interfaceDefinitionHandler;
+
+      public TypeDefinitionVisitor (
+          Action<ITypeDefinitionVisitor, ClassDefinition> classDefinitionHandler,
+          Action<ITypeDefinitionVisitor, InterfaceDefinition> interfaceDefinitionHandler)
+      {
+        ArgumentUtility.CheckNotNull(nameof(classDefinitionHandler), classDefinitionHandler);
+        ArgumentUtility.CheckNotNull(nameof(interfaceDefinitionHandler), interfaceDefinitionHandler);
```

Comment:
Just FYI: for private types, we don't need argument checks (same as with any other private method), except...